File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    const numMap = {}; // Initialize a Hash Map - object in JavaScript
    // Iterate Through the Array
    for(let i = 0; i < nums.length; i++){
        const currentNum = nums[i];
        // Calculate the complement
        const complement = target - currentNum;
        
        // Check if the complement exists in map
        if(numMap.hasOwnProperty(complement)){
            // If found, return the indices
            return [numMap[complement], i];
        }

        // If the complement is not found, add the currentNum and its index to the numMap
        numMap[currentNum] = i;
    }
    return [];
};

File: 0001-two-sum/0001-two-sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        num_to_index = {} # create a empty hash map to store
        
        # Loop through the array - gives indices and current number(num)
        for i, num in enumerate(nums):
            # calculate the complement(number we need to pair with the num to hit target)
            complement = target - num

            # Check hash map -> have we already seen this complement
            if complement in num_to_index:
                # if yes then the current number and earlier complement add up to the target 
                # return indicies
                return [num_to_index[complement], i]
            # if not seen yet, store the current number in the map with its index
            num_to_index[num] = i
        return []